File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-1.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        
        rows, cols = len(heights), len(heights[0])
        directions = [(1,0),(-1,0), (0,1), (0,-1)]
        
        def dfs(r: int, c: int, visited: List[List[bool]]):
            visited[r][c] = True

            for dr, dc in directions:
                nr, nc = r+dr, c + dc
                
                if nr < 0 or nr>=rows or nc<0 or nc>=cols:
                    continue
                
                if visited[nr][nc]:
                    continue
                
                # Reverse flow: move only to cells with
                # height >= current cell                
                if heights[nr][nc] < heights[r][c]:
                    continue

                dfs(nr,nc, visited)


        pacific = [[False]* cols for _ in range(rows)]
        atlantic = [[False]* cols for _ in range(rows)]

        # Pacific top row + left column)
        for c in range(cols):
            dfs(0, c, pacific)

        for r in range(rows):
            dfs(r,0, pacific)
        
        # Atlantic (bottom row + right column)
        for c in range(cols):
            dfs(rows-1, c, atlantic)
        
        for r in range(rows):
            dfs(r, cols-1, atlantic)

        ans = []

        for r in range(rows):
            for c in range(cols):
                if pacific[r][c] and atlantic[r][c]:
                    ans.append([r,c])

        return ans

        



        

File: Data Structures & Algorithms/surrounded-regions/submission-1.cpp
class Solution {
private: 
int rows, cols;
    void dfs(vector<vector<char>>& board, int r, int c)
    {
        if(r<0 || r>=rows || c<0 || c>=cols || board[r][c]!='O')
        return ;

        // if board[r][c]=='O'
        board[r][c] = '#';

        // go all directions
        dfs(board, r + 1, c);
        dfs(board, r - 1, c);
        dfs(board, r, c + 1);
        dfs(board, r, c - 1);        
    }
public:
    void solve(vector<vector<char>>& board) {

        rows = board.size();
        cols = board[0].size();     

        // traverse left and right borders
        for(int r=0;r<rows;r++)
        {
            if(board[r][0]=='O')
             dfs(board,r,0);

            if(board[r][cols-1]=='O')
             dfs(board, r, cols-1);
        }

        // traverse top and bottom borders
        for(int c=0;c<cols;c++)
        {
            if(board[0][c]=='O')
             dfs(board,0,c);

            if(board[rows-1][c]=='O')
             dfs(board,rows-1,c);
        }

        for(int r=0;r<rows;r++)
        {
            for(int c=0;c<cols;c++)
            {
                if(board[r][c]=='O')
                    board[r][c]='X';
                else if(board[r][c]=='#')
                     board[r][c]='O';

            }
        }

    }
};
